harness/determined/launch/deepspeed.py

```diff
@@ -100,29 +104,42 @@ def create_deepspeed_env_file() -> None:
         environ = os.environ.copy()
         for k, v in environ.items():
             if k in INCLUDE:
-                f.write(f"{k}={v}\n")
+                if deepspeed_version >= version.parse("0.6.2"):
+                    f.write(f"{k}={shlex.quote(v)}\n")
+                else:
+                    f.write(f"{k}={v}\n")
 
 
-def create_run_command(master_address: str, hostfile_path: str) -> List[str]:
+def create_run_command(master_address: str, hostfile_path: Optional[str]) -> List[str]:
     # Construct the deepspeed command.
-    deepspeed_process_cmd = [
-        "deepspeed",
-        "-H",
-        hostfile_path,
-        "--master_addr",
-        master_address,
-        "--no_python",
-        "--no_local_rank",
-        "--",
-    ]
+    deepspeed_process_cmd = ["deepspeed"]
+    if hostfile_path is not None:
+        deepspeed_process_cmd += ["-H", hostfile_path]
+    deepspeed_process_cmd += ["--master_addr", master_address, "--no_python", "--no_local_rank"]
+    if deepspeed_version > version.parse("0.6.4"):
+        deepspeed_process_cmd.append("--no_ssh_check")
+    deepspeed_process_cmd.append("--")
     return deepspeed_process_cmd
 
 
+def check_deepspeed_version(multi_machine: bool) -> None:
+    if multi_machine:
+        compatible = True
+        if deepspeed_version > version.parse("0.6.0") and deepspeed_version <= version.parse(
+            "0.6.4"
+        ):
+            compatible = False
+        assert compatible, f"launcher incompatible with deepspeed version {deepspeed.__version__}"
```

Comment:
nit: simply the logic/indentation here:

```python
def check_deepspeed_version(multi_machine: bool) -> None:
    if not multi_machine:
        return
    if deepspeed_version <= version.parse("0.6.0"):
        return
    if deepspeed_version > version.parse("0.6.4"):
        return
    raise ValueError(...)
```

Also, expand the error message to explain which versions _are_ allowed (so the user knows what to do) and explain why the disallowed versions are disallowed (so the user understands why we can't support them).